### DISCUSSION TITLE
useSubscription cached data not updated

### DISCUSSION THREAD

**Discussion by Sytten** [urql-graphql/urql#1721] [Q&A] [answered]
Hi!
I have a problem that I am not sure how to resolve.
I have 1 query and 2 subscriptions:
- Query is relay paginated (with the `relayPagination` extra)  over a list of items
```graphql
{
  todos(first: 100, after: $after) {
    id
    comments {
       id
    }
  }
}
```
- Subscription 1 for new todo:
```graphql
subscription {
   createdTodo {
      id
   }
}
```
- Subscription 2 on new comments:
```graphql
subscription {
   createdComment {
      id
      todo {
         id
      }
   }
}
```

In the vue code I am basically doing:
```typescript
const { data: currentTodos } = useQuery(...)
const { data: newTodos } = useSubscription(...)
return {
    todos: computed(() => [...currentTodos.value, ...newTodos.value])
}
```

When a new comment is created, I update the cache for the given `Todo` and add the new comment.
This works (aka the UI is updated) when the `Todo` is from the query but it doesn't seem to work when the `Todo` is from the subscription.
Is that expected and if so, how should I deal with it?

**Comment by kitten** [comment, marked as answer]
you'll need to use the normalised cache and teach it how to add new comments in your particular schema's structure or alternatively you'll have to handle it in local state 🤔

We have a docs page on cache updates with the normalised cache if that's of interest: https://formidable.com/open-source/urql/docs/graphcache/cache-updates/

**Reply by Sytten** [reply]
Do the subscriptions work like the mutations aka the data is not updated even if you modify the normalized cache?

Because I do custom normalization for the `createdComment` subscription to modify the normalized `Todo` objects but it doesn't seem to propagate to the data `newTodos`.

**Reply by kitten** [reply]
I'm not quite sure what you mean by "custom normalisation", so I can't answer that 😅

It's all explained in our page on normalized caching: https://formidable.com/open-source/urql/docs/graphcache/

what's important to understand is that each entity (so an instance/object shaped after a GraphQL type) is keyed and stored by that key, as far as possible. All these keyed entities are densely connected, so you have entities point at entities sometimes. In a normalised cache this is converted to "link" like a secondary table in databases. In other words entities refer to other entities by jets rather than reference after normalisation.

when a normalised cache reads from cache it needs to start somewhere. That's where in GraphQL we choose to start from the root Query type, since that's guaranteed to always give us stable results, in other words, if no change is made then querying the same data repeatedly gives us the same result.

This last point isn't true for mutations or subscriptions as they can deliver arbitrarily changing data. That means that they can only update existing data and aren't read from cache, they can only write to the cache.

Links are relations, but relations aren't on their own normalized. This means that just because a mutation or a subscription notifies us of new entities, doesn't mean that the normalized cache knows where to put them.

**Reply by Sytten** [reply]
Ok so this confirms what I see for subscriptions. If I update the `Todo` object in the cache to add a comment via the `createdComment` subscription, this doesn't update the `Todo` object stored in the `data` of the `createdTodo` subscription.
So basically my only choice is that the `createdTodo` subscription adds the `Todo` to the `todos` cached query in a random page (are there guidelines for that?) and keep the `createdComment` subscription the same. This means the data only comes from the query and the UI will be updated correctly.
